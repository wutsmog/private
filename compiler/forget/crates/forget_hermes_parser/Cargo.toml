[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }