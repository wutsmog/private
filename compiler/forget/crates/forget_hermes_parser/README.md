# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.