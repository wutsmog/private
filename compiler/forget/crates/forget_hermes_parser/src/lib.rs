mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
