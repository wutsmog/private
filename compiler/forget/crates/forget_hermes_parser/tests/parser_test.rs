use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
