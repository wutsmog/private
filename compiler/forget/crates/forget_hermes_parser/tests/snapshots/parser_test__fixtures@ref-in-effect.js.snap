---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/ref-in-effect.js
---
Input:
function Component(props) {
  const ref = useRef(null);
  const onChange = (e) => {
    const newValue = e.target.value ?? ref.current;
    ref.current = newValue;
  };
  useEffect(() => {
    console.log(ref.current);
  });
  return <Foo onChange={onChange} />;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 19,
            "end": 24
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "ref",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 36,
                    "end": 39
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "Identifier",
                    "name": "useRef",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 42,
                      "end": 48
                    }
                  },
                  "arguments": [
                    {
                      "type": "NullLiteral",
                      "loc": null,
                      "range": {
                        "start": 49,
                        "end": 53
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 42,
                    "end": 54
                  }
                },
                "loc": null,
                "range": {
                  "start": 36,
                  "end": 54
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 30,
              "end": 55
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "onChange",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 64,
                    "end": 72
                  }
                },
                "init": {
                  "type": "ArrowFunctionExpression",
                  "id": null,
                  "params": [
                    {
                      "type": "Identifier",
                      "name": "e",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 76,
                        "end": 77
                      }
                    }
                  ],
                  "body": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "VariableDeclaration",
                        "kind": "const",
                        "declarations": [
                          {
                            "type": "VariableDeclarator",
                            "id": {
                              "type": "Identifier",
                              "name": "newValue",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 94,
                                "end": 102
                              }
                            },
                            "init": {
                              "type": "LogicalExpression",
                              "operator": "??",
                              "left": {
                                "type": "MemberExpression",
                                "object": {
                                  "type": "MemberExpression",
                                  "object": {
                                    "type": "Identifier",
                                    "name": "e",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": {
                                      "start": 105,
                                      "end": 106
                                    }
                                  },
                                  "property": {
                                    "type": "Identifier",
                                    "name": "target",
                                    "typeAnnotation": null,
                                    "loc": null,
                                    "range": {
                                      "start": 107,
                                      "end": 113
                                    }
                                  },
                                  "computed": false,
                                  "loc": null,
                                  "range": {
                                    "start": 105,
                                    "end": 113
                                  }
                                },
                                "property": {
                                  "type": "Identifier",
                                  "name": "value",
                                  "typeAnnotation": null,
                                  "loc": null,
                                  "range": {
                                    "start": 114,
                                    "end": 119
                                  }
                                },
                                "computed": false,
                                "loc": null,
                                "range": {
                                  "start": 105,
                                  "end": 119
                                }
                              },
                              "right": {
                                "type": "MemberExpression",
                                "object": {
                                  "type": "Identifier",
                                  "name": "ref",
                                  "typeAnnotation": null,
                                  "loc": null,
                                  "range": {
                                    "start": 123,
                                    "end": 126
                                  }
                                },
                                "property": {
                                  "type": "Identifier",
                                  "name": "current",
                                  "typeAnnotation": null,
                                  "loc": null,
                                  "range": {
                                    "start": 127,
                                    "end": 134
                                  }
                                },
                                "computed": false,
                                "loc": null,
                                "range": {
                                  "start": 123,
                                  "end": 134
                                }
                              },
                              "loc": null,
                              "range": {
                                "start": 105,
                                "end": 134
                              }
                            },
                            "loc": null,
                            "range": {
                              "start": 94,
                              "end": 134
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 88,
                          "end": 135
                        }
                      },
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "AssignmentExpression",
                          "operator": "=",
                          "left": {
                            "type": "MemberExpression",
                            "object": {
                              "type": "Identifier",
                              "name": "ref",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 140,
                                "end": 143
                              }
                            },
                            "property": {
                              "type": "Identifier",
                              "name": "current",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 144,
                                "end": 151
                              }
                            },
                            "computed": false,
                            "loc": null,
                            "range": {
                              "start": 140,
                              "end": 151
                            }
                          },
                          "right": {
                            "type": "Identifier",
                            "name": "newValue",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 154,
                              "end": 162
                            }
                          },
                          "loc": null,
                          "range": {
                            "start": 140,
                            "end": 162
                          }
                        },
                        "directive": null,
                        "loc": null,
                        "range": {
                          "start": 140,
                          "end": 163
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 82,
                      "end": 167
                    }
                  },
                  "generator": false,
                  "async": false,
                  "loc": null,
                  "range": {
                    "start": 75,
                    "end": 167
                  },
                  "expression": false,
                  "loc": null,
                  "range": {
                    "start": 75,
                    "end": 167
                  }
                },
                "loc": null,
                "range": {
                  "start": 64,
                  "end": 167
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 58,
              "end": 168
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "useEffect",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 171,
                  "end": 180
                }
              },
              "arguments": [
                {
                  "type": "ArrowFunctionExpression",
                  "id": null,
                  "params": [],
                  "body": {
                    "type": "BlockStatement",
                    "body": [
                      {
                        "type": "ExpressionStatement",
                        "expression": {
                          "type": "CallExpression",
                          "callee": {
                            "type": "MemberExpression",
                            "object": {
                              "type": "Identifier",
                              "name": "console",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 193,
                                "end": 200
                              }
                            },
                            "property": {
                              "type": "Identifier",
                              "name": "log",
                              "typeAnnotation": null,
                              "loc": null,
                              "range": {
                                "start": 201,
                                "end": 204
                              }
                            },
                            "computed": false,
                            "loc": null,
                            "range": {
                              "start": 193,
                              "end": 204
                            }
                          },
                          "arguments": [
                            {
                              "type": "MemberExpression",
                              "object": {
                                "type": "Identifier",
                                "name": "ref",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 205,
                                  "end": 208
                                }
                              },
                              "property": {
                                "type": "Identifier",
                                "name": "current",
                                "typeAnnotation": null,
                                "loc": null,
                                "range": {
                                  "start": 209,
                                  "end": 216
                                }
                              },
                              "computed": false,
                              "loc": null,
                              "range": {
                                "start": 205,
                                "end": 216
                              }
                            }
                          ],
                          "loc": null,
                          "range": {
                            "start": 193,
                            "end": 217
                          }
                        },
                        "directive": null,
                        "loc": null,
                        "range": {
                          "start": 193,
                          "end": 218
                        }
                      }
                    ],
                    "loc": null,
                    "range": {
                      "start": 187,
                      "end": 222
                    }
                  },
                  "generator": false,
                  "async": false,
                  "loc": null,
                  "range": {
                    "start": 181,
                    "end": 222
                  },
                  "expression": false,
                  "loc": null,
                  "range": {
                    "start": 181,
                    "end": 222
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 171,
                "end": 223
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 171,
              "end": 224
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "Foo",
                  "loc": null,
                  "range": {
                    "start": 235,
                    "end": 238
                  }
                },
                "attributes": [
                  {
                    "type": "JSXAttribute",
                    "name": {
                      "type": "JSXIdentifier",
                      "name": "onChange",
                      "loc": null,
                      "range": {
                        "start": 239,
                        "end": 247
                      }
                    },
                    "value": {
                      "type": "JSXExpressionContainer",
                      "expression": {
                        "type": "Identifier",
                        "name": "onChange",
                        "typeAnnotation": null,
                        "loc": null,
                        "range": {
                          "start": 249,
                          "end": 257
                        }
                      },
                      "loc": null,
                      "range": {
                        "start": 248,
                        "end": 258
                      }
                    },
                    "loc": null,
                    "range": {
                      "start": 239,
                      "end": 258
                    }
                  }
                ],
                "selfClosing": true,
                "loc": null,
                "range": {
                  "start": 234,
                  "end": 261
                }
              },
              "children": [],
              "closingElement": null,
              "loc": null,
              "range": {
                "start": 234,
                "end": 261
              }
            },
            "loc": null,
            "range": {
              "start": 227,
              "end": 262
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 26,
          "end": 264
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 264
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 264
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 264
  }
}
